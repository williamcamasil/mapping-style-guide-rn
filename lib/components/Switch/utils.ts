import { AppThemeType } from '../../theme';

export function getDisabledTouchabledStyle(theme: AppThemeType, disabled?: boolean) {
  if (!disabled) return null;

  return {
    backgroundColor: theme.colors.neutralGray200,
  };
}

export function getBackgroundColorEllipse(theme: AppThemeType, disabled?: boolean) {
  if (disabled) return theme.colors.neutralGray300;

  return theme.colors.neutralWhite;
}
