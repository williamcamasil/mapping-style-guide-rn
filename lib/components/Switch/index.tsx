import React, { useCallback, useRef } from 'react';
import {
  Animated, StyleProp, TouchableOpacity, ViewStyle,
} from 'react-native';

import Color from 'color';
import { useDidMountAndUpdate } from 'mapping-context-rn';

import { useViewStyles } from '../../hooks';
import { AppThemeType, withTheme } from '../../theme';
import { wrapField } from '../Form';
import { getBackgroundColorEllipse, getDisabledTouchabledStyle } from './utils';

export type SwitchPropsType = {
  theme: AppThemeType;
  testID?: string;
  onChecked?: (checked: boolean) => void;
  onBlur?: () => void;
  onFocus?: () => void;
  style?: StyleProp<ViewStyle>;
  disabled?: boolean;
  checked?: boolean;
};

const CHECKED_TRANSLATE = 20;
const UNCHECKED_TRANSLATE = 0;

const SPRING_CONFIG: Partial<Animated.SpringAnimationConfig> = {
  speed: 20,
  bounciness: 10,
};

const AnimatedTouchableOpacity = Animated.createAnimatedComponent(TouchableOpacity);

const Switch = ({
  testID,
  style,
  onChecked,
  onFocus,
  onBlur,
  disabled,
  checked,
  theme,
}: SwitchPropsType) => {
  const getNewValue = () => (checked ? CHECKED_TRANSLATE : UNCHECKED_TRANSLATE);

  const animationPosition = useRef(new Animated.Value(getNewValue()));
  const animationColor = useRef(new Animated.Value(getNewValue()));

  useDidMountAndUpdate(() => {
    const animation = Animated.parallel([
      Animated.spring(animationPosition.current, {
        ...SPRING_CONFIG,
        toValue: getNewValue(),
        useNativeDriver: true,
      }),
      Animated.spring(animationColor.current, {
        ...SPRING_CONFIG,
        toValue: getNewValue(),
        useNativeDriver: false,
      }),
    ], {
      stopTogether: true,
    });

    animation.start();

    return () => {
      animation.stop();
    };
  }, [checked]);

  const switchTouchable = useViewStyles(() => {
    const width = 48;
    return [
      {
        width,
        height: 28,
        borderRadius: width / 2,
        padding: 2,
        backgroundColor: animationColor.current.interpolate({
          inputRange: [UNCHECKED_TRANSLATE, CHECKED_TRANSLATE],
          outputRange: [
            Color(theme.colors.neutralGray300).rgb().string(),
            Color(theme.colors.primaryMain).rgb().string(),
          ],
        }),
      },
      getDisabledTouchabledStyle(theme, disabled),
      style,
    ];
  }, [disabled, style, theme]);

  const styleEllipse = useViewStyles(
    () => {
      const size = 24;
      return [
        {
          width: size,
          height: size,
          backgroundColor: getBackgroundColorEllipse(theme, disabled),
          borderRadius: size / 2,
          transform: [{
            translateX: animationPosition.current,
          }],
        },
      ];
    },
    [disabled, theme],
  );

  const handlePress = useCallback(() => {
    onFocus?.();
    onChecked?.(!checked);
    onBlur?.();
  }, [checked, onBlur, onChecked, onFocus]);

  return (
    <AnimatedTouchableOpacity
      testID={testID}
      onPress={handlePress}
      disabled={disabled}
      activeOpacity={theme.opacities.intense}
      style={switchTouchable}
    >
      <Animated.View testID="switch-ellipse-view" style={styleEllipse} />
    </AnimatedTouchableOpacity>
  );
};

Switch.Field = withTheme(wrapField<SwitchPropsType, boolean | undefined>((props, fieldProps) => {
  const {
    value, onChange, onBlur,
    onFocus,
  } = fieldProps;

  return (
    <Switch
      {...props}
      checked={value || false}
      onChecked={onChange}
      onBlur={onBlur}
      onFocus={onFocus}
    />
  );
}));

export default withTheme(Switch);
