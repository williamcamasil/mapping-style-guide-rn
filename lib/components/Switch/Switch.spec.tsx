import React from 'react';
import { StyleSheet } from 'react-native';

import {
  fireEvent, render,
} from '@testing-library/react-native';
import { createForm } from 'final-form';
import { Form } from 'react-final-form';

import Switch from '.';
import DefaultTheme from '../../theme/default';

describe('Switch snapshot', () => {
  it('switch true', () => {
    const tree = render(
      <Switch checked onChecked={jest.fn} />,
    ).toJSON();

    expect(tree).toMatchSnapshot();
  });

  it('switch false', () => {
    const tree = render(
      <Switch checked={false} onChecked={jest.fn} />,
    ).toJSON();

    expect(tree).toMatchSnapshot();
  });

  it('form', () => {
    const result = render(
      <Form
        onSubmit={jest.fn()}
      >
        {() => (
          <Switch.Field
            name="switchField"
          />
        )}
      </Form>,
    );

    const tree = result.toJSON();

    expect(tree).toMatchSnapshot();
  });
});

describe('Switch', () => {
  it('renders correctly', () => {
    render(
      <Switch checked={false} onChecked={jest.fn} />,
    );
  });

  it('should verify if switch is select false', () => {
    const result = render(
      <Switch testID="switch-id" checked={false} onChecked={jest.fn} />,
    );

    const switchComponent = result.getByTestId('switch-id');
    const switchStyle = StyleSheet.flatten(switchComponent.props.style);

    const ellipseComponent = result.getByTestId('switch-ellipse-view');
    const ellipseStyle = StyleSheet.flatten(ellipseComponent.props.style);

    const valueTransform = [{ translateX: 0 }];

    expect(switchStyle).toHaveProperty('backgroundColor', 'rgba(225, 226, 228, 1)');
    expect(ellipseStyle).toHaveProperty('backgroundColor', DefaultTheme.colors.neutralWhite);
    expect(ellipseStyle).toHaveProperty('transform', valueTransform);

  });

  it('should verify if switch is select true', () => {
    const result = render(
      <Switch testID="switch-id" checked onChecked={jest.fn} />,
    );

    const switchComponent = result.getByTestId('switch-id');
    const switchStyle = StyleSheet.flatten(switchComponent.props.style);

    const ellipseComponent = result.getByTestId('switch-ellipse-view');
    const ellipseStyle = StyleSheet.flatten(ellipseComponent.props.style);

    const valueTransform = [{ translateX: 20 }];

    expect(switchStyle).toHaveProperty('backgroundColor', 'rgba(1, 49, 136, 1)');
    expect(ellipseStyle).toHaveProperty('backgroundColor', DefaultTheme.colors.neutralWhite);
    expect(ellipseStyle).toHaveProperty('transform', valueTransform);
  });

  it('should verify if switch is disabled', () => {
    const handlePress = jest.fn();

    const result = render(
      <Switch testID="switch-id" checked disabled onChecked={handlePress} />,
    );

    const switchComponent = result.getByTestId('switch-id');
    const switchStyle = StyleSheet.flatten(switchComponent.props.style);

    const ellipseComponent = result.getByTestId('switch-ellipse-view');
    const ellipseStyle = StyleSheet.flatten(ellipseComponent.props.style);

    fireEvent.press(switchComponent);

    expect(handlePress).not.toBeCalled();

    expect(switchStyle).toHaveProperty('backgroundColor', DefaultTheme.colors.neutralGray200);
    expect(ellipseStyle).toHaveProperty('backgroundColor', DefaultTheme.colors.neutralGray300);
  });

  it('should verify click in switch and call event handlePress', () => {
    const handlePress = jest.fn();

    const result = render(
      <Switch testID="switch-id" checked onChecked={handlePress} />,
    );

    const switchComponent = result.getByTestId('switch-id');

    fireEvent.press(switchComponent);

    expect(handlePress).toBeCalled();
  });
});

describe('Switch.Field', () => {
  it('form valid', () => {
    const form = createForm({
      onSubmit: jest.fn(),
    });

    const result = render(
      <Form
        onSubmit={jest.fn()}
        form={form}
      >
        {() => (
          <Switch.Field
            name="switchname"
            required
            validate={checked => (checked ? undefined : 'Erro')}
            testID="switch-field"
          />
        )}
      </Form>,
    );

    const checkbox = result.getByTestId('switch-field');

    fireEvent.press(checkbox);

    expect(form.getState().valid).toBeTruthy();
  });
});
