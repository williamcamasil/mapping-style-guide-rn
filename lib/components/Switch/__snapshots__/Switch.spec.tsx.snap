// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Switch snapshot form 1`] = `
<View
  accessibilityState={
    {
      "busy": undefined,
      "checked": undefined,
      "disabled": undefined,
      "expanded": undefined,
      "selected": undefined,
    }
  }
  accessibilityValue={
    {
      "max": undefined,
      "min": undefined,
      "now": undefined,
      "text": undefined,
    }
  }
  accessible={true}
  collapsable={false}
  focusable={true}
  forwardedRef={[Function]}
  onClick={[Function]}
  onResponderGrant={[Function]}
  onResponderMove={[Function]}
  onResponderRelease={[Function]}
  onResponderTerminate={[Function]}
  onResponderTerminationRequest={[Function]}
  onStartShouldSetResponder={[Function]}
  style={
    {
      "backgroundColor": "rgba(225, 226, 228, 1)",
      "borderRadius": 24,
      "height": 28,
      "opacity": 1,
      "padding": 2,
      "width": 48,
    }
  }
>
  <View
    collapsable={false}
    style={
      {
        "backgroundColor": "#FFFFFF",
        "borderRadius": 12,
        "height": 24,
        "transform": [
          {
            "translateX": 0,
          },
        ],
        "width": 24,
      }
    }
    testID="switch-ellipse-view"
  />
</View>
`;

exports[`Switch snapshot switch false 1`] = `
<View
  accessibilityState={
    {
      "busy": undefined,
      "checked": undefined,
      "disabled": undefined,
      "expanded": undefined,
      "selected": undefined,
    }
  }
  accessibilityValue={
    {
      "max": undefined,
      "min": undefined,
      "now": undefined,
      "text": undefined,
    }
  }
  accessible={true}
  collapsable={false}
  focusable={true}
  forwardedRef={[Function]}
  onClick={[Function]}
  onResponderGrant={[Function]}
  onResponderMove={[Function]}
  onResponderRelease={[Function]}
  onResponderTerminate={[Function]}
  onResponderTerminationRequest={[Function]}
  onStartShouldSetResponder={[Function]}
  style={
    {
      "backgroundColor": "rgba(225, 226, 228, 1)",
      "borderRadius": 24,
      "height": 28,
      "opacity": 1,
      "padding": 2,
      "width": 48,
    }
  }
>
  <View
    collapsable={false}
    style={
      {
        "backgroundColor": "#FFFFFF",
        "borderRadius": 12,
        "height": 24,
        "transform": [
          {
            "translateX": 0,
          },
        ],
        "width": 24,
      }
    }
    testID="switch-ellipse-view"
  />
</View>
`;

exports[`Switch snapshot switch true 1`] = `
<View
  accessibilityState={
    {
      "busy": undefined,
      "checked": undefined,
      "disabled": undefined,
      "expanded": undefined,
      "selected": undefined,
    }
  }
  accessibilityValue={
    {
      "max": undefined,
      "min": undefined,
      "now": undefined,
      "text": undefined,
    }
  }
  accessible={true}
  collapsable={false}
  focusable={true}
  forwardedRef={[Function]}
  onClick={[Function]}
  onResponderGrant={[Function]}
  onResponderMove={[Function]}
  onResponderRelease={[Function]}
  onResponderTerminate={[Function]}
  onResponderTerminationRequest={[Function]}
  onStartShouldSetResponder={[Function]}
  style={
    {
      "backgroundColor": "rgba(233, 88, 20, 1)",
      "borderRadius": 24,
      "height": 28,
      "opacity": 1,
      "padding": 2,
      "width": 48,
    }
  }
>
  <View
    collapsable={false}
    style={
      {
        "backgroundColor": "#FFFFFF",
        "borderRadius": 12,
        "height": 24,
        "transform": [
          {
            "translateX": 20,
          },
        ],
        "width": 24,
      }
    }
    testID="switch-ellipse-view"
  />
</View>
`;
